src/market/cache.rs

```diff
@@ -0,0 +1,295 @@
+use async_trait::async_trait;
+use chrono::{DateTime, Duration, Utc};
+pub use implementation::Cache;
+use std::fmt::Debug;
+
+#[derive(Debug)]
+pub struct Cached<T> {
+    pub record: T,
+    pub expires: DateTime<Utc>,
+}
+
+impl<R> Cached<R> {
+    pub fn new(record: R, expire_after: Duration) -> Self {
+        Self {
+            record,
+            expires: Utc::now() + expire_after,
+        }
+    }
+
+    pub fn is_expired(&self) -> bool {
+        Utc::now() >= self.expires
+    }
+
+    pub fn is_valid(&self) -> bool {
+        !self.is_expired()
+    }
+}
+
+#[async_trait(?Send)]
+pub trait ClientLike<K> {
+    type Output: Send + 'static;
+
+    async fn get_fresh<'a>(&self, key: K) -> Self::Output
```

Comment:
I just couldn't figure out a way for `K` to be able to support owned and reference values... I've tried mimicking `DashMap` and `HashMap`, but I got beaten by `async` and the `async_trait`